torch/distributed/_sharded_tensor/api.py

```diff
@@ -750,3 +796,61 @@ def _create_tensor_from_params(*size, local_device, tensor_init_params: TensorIn
                           device=local_device, )
     else:
         raise ValueError(f'Unsupported create_op: {tensor_init_params.create_op}')
+
+
+class PartialTensor(object):
+    """
+    PartialTensor is an abstraction to represent Tensors that need
+    aggregation across multiple devices and multiple processes.
+
+    PartialTensor is initialized in an SPMD like fashion where each rank
+    initializes the PartialTensor. The PartialTensor object on each rank
+    then only stores the local partial shard and provides global metadata
+    for all the shards.
+
+    PartialTensor doesn't provide any Tensor like operations but is a
+    wrapper providing the Tensor representing the local partial shard
+    and the global metadata.
+
+    Using these, users can apply custom distributed sharded computations
+    on top of this primitive. The local shards are all initialized using
+    the create_op specified by tensor_init_params.create_op, e.g.,
+    torch.ones, or torch.empty.
+
+    Args:
+        local_partial_results (List(Tensor)):
+            Partial results are stored across ranks and in a form of a list.
+    """
+
+    def __init__(self, local_partial_results):
```

Comment:
We need an additional enum here for type of aggregation needed.

Reply:
sure, let me update the code to reflect the design.

Reply:
I think we can put that enum into resharding spec. LKM if this sounds make sense to you or not.

Reply:
Hmm I don't think the type of aggregation fits in the resharding spec. ReshardingSpec probably only mentions how to reshard the data but shouldn't care about how to aggregate the data into a single Tensor.

Reply:
I see. I think what you mean here because the aggregation is kind of exclusive to the partial tensor. Another idea I have is that, can we directly use the enum `ReduceOp` instead of creating a new one ourselves?